proxy/proxy.go

```diff
@@ -0,0 +1,174 @@
+package proxy
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gorilla/mux"
+)
+
+const (
+	watchdogPort       = "8080"
+	defaultContentType = "text/plain"
+)
+
+// BaseURLResolver URL resolver for proxy requests
+//
+// The FaaS provider implementation is responsible for providing the resolver function implementation.
+// BaseURLResolver.Resolve will receive the function name and should return the base Address of the
+// function service.
+type BaseURLResolver interface {
+	Resolve(functionName string) (string, error)
```

Comment:
Is the convention in go to use `string` for URL or `https://golang.org/pkg/net/url` ?

Reply:
I believe this is porting rather than refactoring the existing code, correct @LucasRoesler ?

Reply:
There is a slight refactoring because it needs to expose a way for the provider to customize the URL, but this is similar in style to the original proxy code

Reply:
(This code in the original faas repo appears to use a URL though)

https://github.com/openfaas/faas/blob/master/gateway/handlers/forwarding_proxy.go#L39

Reply:
@LucasRoesler what was the reason for going for string over URL? It looked like the original code used URL but I might be off on this.

Reply:
After additional discussion I am going to update this interface so taht a URL is returned. 

```golang
Resolve(functionName string) (url.URL, error)
```